ee/localserver/krypto-ec-middleware.go

```diff
@@ -459,3 +366,369 @@ func (bhr *bufferedHttpResponse) WriteHeader(code int) {
+	if err != nil {
+		e.slogger.Log(ctx, slog.LevelError,
+			"error parsing presence detection interval",
+			"err", err,
+		)
+		return
+	}
+
+	// On MacOS presence detection appears as "Kolide is trying to {reason}."
+	reason := "authenticate"
+
+	reasonKey := kolidePresenceDetectionReasonMacosHeaderKey
+	if runtime.GOOS == "windows" {
+		// On windows presence detection text is expected to be a full sentence
+		reason = "Kolide is requesting authentication"
+		reasonKey = kolidePresenceDetectionReasonWindowsHeaderKey
+	}
+
+	if reasonStr, ok := cmdReq.CallbackHeaders[reasonKey]; ok && len(reasonStr) > 0 && len(reasonStr[0]) > 0 {
+		reason = reasonStr[0]
+	}
+
+	if !e.presenceDetectionLock.TryLock() {
+		e.slogger.Log(ctx, slog.LevelInfo,
+			"dropping presence detection callback, already in progress",
+		)
+		return
+	}
+	defer e.presenceDetectionLock.Unlock()
+
+	type detectionResult struct {
+		durationSinceLastDetection time.Duration
+		err                        error
+	}
+
+	presenceDetectionStartTime := time.Now()
+
+	detectionDoneChan := make(chan *detectionResult)
+
+	// kick of presence detection
+	gowrapper.Go(ctx, e.slogger, func() {
+		durationSinceLastDetection, err := e.presenceDetector.DetectPresence(reason, detectionIntervalDuration)
+		if err != nil {
+			e.slogger.Log(ctx, slog.LevelInfo,
+				"presence_detection",
+				"reason", reason,
+				"interval", detectionIntervalDuration,
+				"duration_since_last_detection", durationSinceLastDetection,
+				"err", err,
+			)
+		}
+
+		detectionDoneChan <- &detectionResult{
+			durationSinceLastDetection: durationSinceLastDetection,
+			err:                        err,
+		}
+	})
+
+	statusUpdateTicker := time.NewTicker(e.presenceDetectionStatusUpdateInterval)
+	defer statusUpdateTicker.Stop()
+
+	headers := map[string][]string{
+		kolideArchHeaderKey: {runtime.GOARCH},
+		kolideOsHeaderKey:   {runtime.GOOS},
+	}
+
+	kolideSessionId, ok := cmdReq.CallbackHeaders[kolideSessionIdHeaderKey]
+	if ok {
+		headers[kolideSessionIdHeaderKey] = kolideSessionId
+	}
+
+	type presenceDetectionMessages string
+
+	const (
+		presenceDetectionCompleted presenceDetectionMessages = "presence_detection_complete"
+		presenceDetectionTimedOut  presenceDetectionMessages = "presence_detection_timed_out"
+		presenceDetectionWaiting   presenceDetectionMessages = "presence_detection_waiting"
+		presenceDetectionError     presenceDetectionMessages = "presence_detection_error"
+	)
+
+	var finalPresenceDetectionResult *detectionResult
+	hasPresenceDetectionTimedout := false
+
+	// send status updates to k2 on an interval until presence detection completes
+	for {
+
+		// Build status update for k2
+		callbackBody := map[string]string{
+			"elapsed_time": time.Since(presenceDetectionStartTime).String(),
+		}
+
+		switch {
+
+		// completed with error
+		case finalPresenceDetectionResult != nil && finalPresenceDetectionResult.err != nil:
+			e.slogger.DebugContext(ctx,
```

Comment:
DebugContext => Log, also maybe level Info or Warn so we get these logs shipped up?

Reply:
done, set it to warn